tests/test_rpub.py

```diff
@@ -0,0 +1,837 @@
+        q.insert(key, value)
+        expected += 1
+
+    def remove():
+        nonlocal expected
+        result = q.pop_highest_priority()
+        if result is not None:
+            expected -= 1
+
+    add(1, "a")
+    assert len(q) == expected
+    add(2, "b")
+    assert len(q) == expected
+    remove()
+    assert len(q) == expected
+    add(0, "c")
+    assert len(q) == expected
+    remove()
+    assert len(q) == expected
+    remove()
+    assert len(q) == expected
+    remove()
+    assert len(q) == expected
+
+
+def test_queue_contains():
+    q = RAPriorityQueue()
+    assert "foo" not in q
+    q.insert("foo", "bar")
+    assert "foo" in q
+    q.pop_highest_priority()
+    assert "foo" not in q
+    q.insert("baz", "quux")
+    q.insert("xen", "hom")
+    assert "baz" in q
+    assert "xen" in q
+
+
+def test_queue_insert_setitem():
+    q = RAPriorityQueue()
+    assert "foo" not in q
+    q.insert("foo", "bar")
+    assert "foo" in q
+    assert q["foo"] == "bar"
+    q.insert("foo", "baz")
+    assert q["foo"] == "baz"
+    q["foo"] = "quux"
+    assert q["foo"] == "quux"
+    q["bar"] = "xen"
+    assert q["bar"] == "xen"
+
+
+def test_queue_getitem():
+    q = RAPriorityQueue()
+    with pytest.raises(KeyError):
+        q["foo"]
+    q.insert("foo", "bar")
+    assert q["foo"] == "bar"
+    q.insert("baz", "quux")
+    assert q["foo"] == "bar"
+    assert q["baz"] == "quux"
+    q.pop_highest_priority()
+    assert q["foo"] == "bar"
+    with pytest.raises(KeyError):
+        q["baz"]
+
+
+def test_queue_pop():
+    q = RAPriorityQueue()
+    assert q.pop_highest_priority() is None
+    # if there is only one item, pop must remove it
+    q.insert(1, "foo")
+    assert q.pop_highest_priority() == (1, "foo")
+    # multiple items must be removed in priority order
+    q.insert(2, "bar")
+    q.insert(3, "baz")
+    q.insert(1, "foo")
+    assert q.pop_highest_priority() == (1, "foo")
+    assert q.pop_highest_priority() == (2, "bar")
+    assert q.pop_highest_priority() == (3, "baz")
+
+
+def test_queue_remove_del():
+    q = RAPriorityQueue()
+    with pytest.raises(KeyError):
+        q.remove(1)
+    q.insert(2, "bar")
+    q.insert(3, "baz")
+    q.insert(1, "foo")
+    q.remove(2)
+    assert 2 not in q
+    assert q.pop_highest_priority() == (1, "foo")
+    assert q.pop_highest_priority() == (3, "baz")
+    with pytest.raises(KeyError):
+        del q[1]
+    q.insert(2, "bar")
+    q.insert(3, "baz")
+    q.insert(1, "foo")
+    del q[1]
+    assert 1 not in q
+    assert q.pop_highest_priority() == (2, "bar")
+    assert q.pop_highest_priority() == (3, "baz")
+
+
+########################################
+
+
+def test_journal_construct_no_file(tmpdir):
+    journal_path = tmpdir.join("journal")
+    j = PublicationJournal(journal_path)
+    assert not j.has_messages_to_send()
+    assert not j.has_messages_in_flight()
+    assert j.get_next_message_to_send() is None
+    assert os.path.exists(journal_path)
```

Comment:
The `PublicationJournal` constructor creates the journal file, so the check is correct, if not very useful/important. It could be argued that this isn't necessary, so we shouldn't test for it, but I would not want to enforce the reverse, as it would complicate the journal implementation to have to open the journal for writing lazily.